std/socket.zig

```diff
@@ -0,0 +1,678 @@
+}
+
+// add timeouts
+
+pub const Socket = struct {
+    fd: SocketFd,
+    domain: Domain,
+    socket_type: SocketType,
+    protocol: Protocol,
+
+    fn initSocket(fd: SocketFd, domain: Domain, socket_type: SocketType, protocol: Protocol) Socket {
+        return Socket{
+            .fd = @intCast(SocketFd, rc),
+            .domain = domain,
+            .socket_type = socket_type,
+            .protocol = protocol,
+        };
+    }
+
+    pub fn new(domain: Domain, socket_type: SocketType, protocol: Protocol) SocketError!Socket {
+        if (is_posix) {
+            const rc = system.socket(domain.toInt(u32), socket_type.toInt(u32), try protocol.toInt(u32));
+            const err = system.getErrno(rc);
+            switch (err) {
+                0 => return initSocket(@intCast(SocketFd, rc), domain, socket_type, protocol),
+                system.EACCES => return SocketError.PermissionDenied,
+                system.EAFNOSUPPORT => return SocketError.AddressFamilyNotSupported,
+                system.EINVAL => return SocketError.ProtocolFamilyNotAvailable,
+                system.EMFILE => return SocketError.ProcessFdQuotaExceeded,
+                system.ENFILE => return SocketError.system.emFdQuotaExceeded,
+                system.ENOBUFS, system.ENOMEM => return SocketError.system.emResources,
+                system.EPROTONOSUPPORT => return SocketError.ProtocolNotSupported,
+                else => return unexpectedError(err),
+            }
+        } else {
+            const rc = system.socket(domain.toInt(c_int), sock_type.toInt(c_int), try protocol.toInt(c_int));
+            const err = system.WSAGetLastError();
+            // TODO check for TOO_MANY_OPEN_FILES?
+            switch (err) {
+                0 => return initSocket(rc, domain, socket_type, protocol),
+                system.ERROR.WSAEACCES => return SocketError.PermissionDenied,
+                system.ERROR.WSAEAFNOSUPPORT => return SocketError.AddressFamilyNotSupported,
+                system.ERROR.WSAEINVAL => return SocketError.ProtocolFamilyNotAvailable,
+                system.ERROR.WSAEMFILE => return SocketError.ProcessFdQuotaExceeded,
+                system.ERROR.WSAENOBUFS => return SocketError.system.emResources,
+                system.ERROR.WSAEPROTONOSUPPORT, system.ERROR.WSAEPROTOTYPE, system.ERROR.WSAESOCKTNOSUPPORT => return SocketError.ProtocolNotSupported,
+                else => return unexpectedError(@intCast(u32, err)),
+            }
+        }
+    }
+
+    pub fn tcp(domain: Domain) SocketError!Socket {
+        return Socket.new(domain, SocketType.Stream, Protocol.TCP);
+    }
+
+    pub fn udp(domain: Domain) SocketError!Socket {
+        return Socket.new(domain, SocketType.DataGram, Protocol.UDP);
+    }
+
+    pub fn unix(socket_type: SocketType) SocketError!Socket {
+        return Socket.new(Domain.Unix, socket_type, Protocol.Unix);
+    }
+
+    pub fn bind(self: Socket, addr: *const net.Address) BindError!void {
```

Comment:
Most of these member functions are just forwarding to "free functions" under-the-hood and passing the socket handle as the first parameter.  What's the reason for changing the underlying ABI and making them member functions?  I would have thought the `Socket` type would just be an alias for the socket handle and that these functions would be free functions that accept `Socket` as the first parameter and forward the call to the platform-specific function and convert error codes.

I would start by mimicking the original API and only change it if it provides some benefit.  I'm not seeing the benefit here.

Reply:
this is the sort of pattern that the stdlib adheres to for `std.fs.File` API so i thought it would make sense to copy it. also did you mean API instead of ABI? making these functions namespaced should change the ABI...

Reply:
One big difference here is that the `File` struct only has one field, the handle.  The `Socket` type would be less problematic if it only had the one field like the `File` type.  But since you've removed the original `socket` functions, now if I want to call one I have to allocate a struct with 4 fields on the stack, then call a member function on that struct which will only be using the handle field anyway.

Reply:
Note that socket handles will be passed back and forth between the OS as normal handles (i.e. usize or c_int or something).  For example, if you're handling sockets from a `select` call or `poll` or `epoll` or `iocp`.  With this API, anytime you receive a socket handle from the OS, you have to reconstruct a `Socket` struct with 4 fields just to call functions on it.